unifi/service.go

```diff
@@ -0,0 +1,260 @@
+	unifToken      string
+	nvrAddr        string
+	httpClientFunc func() *http.Client
+}
+
+// Validate validates the Unifi discovery service configuration.
+func (cfg *Config) Validate(_ string) ([]string, []string, error) {
+	if cfg.NVRAddress == "" {
+		return nil, nil, errors.New("nvr_address is required")
+	}
+	if cfg.UnifiToken == "" {
+		return nil, nil, errors.New("unifi_token is required")
+	}
+	return nil, nil, nil
+}
+
+func init() {
+	resource.RegisterService(
+		discovery.API,
+		Model,
+		resource.Registration[discovery.Service, *Config]{
+			Constructor: newUnifiDiscovery,
+		})
+}
+
+// NewUnifiDiscovery creates a new Unifi discovery service (exported for testing).
+func NewUnifiDiscovery(ctx context.Context, deps resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (discovery.Service, error) {
+	return newUnifiDiscovery(ctx, deps, conf, logger)
+}
+
+func newUnifiDiscovery(_ context.Context, _ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (discovery.Service, error) {
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	dis := &unifiDiscovery{
+		Named:          conf.ResourceName().AsNamed(),
+		unifToken:      cfg.UnifiToken,
+		nvrAddr:        cfg.NVRAddress,
+		logger:         logger,
+		httpClientFunc: defaultHTTPClient,
+	}
+
+	return dis, nil
+}
+
+func (dis *unifiDiscovery) DiscoverResources(ctx context.Context, _ map[string]any) ([]resource.Config, error) {
+	// Get list of cameras from NVR
+	cameras, err := dis.getCameras(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cameras: %w", err)
+	}
+
+	dis.logger.Infof("Found %d cameras on NVR %s", len(cameras), dis.nvrAddr)
+
+	var configs []resource.Config
+
+	// For each camera, get the RTSPS stream URL and convert to RTSP
+	for _, cam := range cameras {
+		rtspURL, err := dis.getRTSPStream(ctx, cam.ID)
+		if err != nil {
+			dis.logger.Warnf("Failed to get RTSP stream for camera %s (%s): %v", cam.Name, cam.ID, err)
+			continue
+		}
+
+		if rtspURL == "" {
+			dis.logger.Warnf("No RTSP stream available for camera %s (%s)", cam.Name, cam.ID)
+			continue
+		}
+
+		dis.logger.Infof("Camera %s: %s", cam.Name, rtspURL)
+
+		// Create a camera config for this discovered camera
+		cfg := resource.Config{
+			API:   camera.API,
+			Model: viamrtsp.ModelAgnostic,
+			Name:  sanitizeName(cam.Name, cam.ID),
+			Attributes: map[string]any{
+				"rtsp_address": rtspURL,
+			},
+		}
+		configs = append(configs, cfg)
+	}
+
+	return configs, nil
+}
+
+func (dis *unifiDiscovery) getCameras(ctx context.Context) ([]unifiCamera, error) {
+	url := fmt.Sprintf("https://%s/proxy/protect/integration/v1/cameras", dis.nvrAddr)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("X-Api-Key", dis.unifToken)
+	req.Header.Set("Accept", "application/json")
+
+	resp, err := dis.httpClient().Do(req)
```

Comment:
I could be reading this wrong, but does this mean we spin up a new `httpClient` every call? If that is the case, we should probably use the singleton pattern (1 client per 1 resource configured)

Can implement this using: https://pkg.go.dev/sync#Once.Do

Reply:
Cool, `once.Do(f)` pattern looks nice.

I was thinking that since DiscoverResources is called infrequently it would make sense to just spin up the http client each call, but, realizing http clients do not hold open ports or anything like normal TCP so there is no downside to keeping around.

Seems more readable to simply put http creation in the constructor and add to struct, changing

Reply:
Creating http client once in the `newUnifiDiscovery` constructor now. [commit](https://github.com/viam-modules/viamrtsp/pull/166/commits/85c56b4c34917eba1359492b50a2700507e5a6f1)

Reply:
sounds good thanks